### ISSUE TITLE
Messages take long time to appear

### ISSUE THREAD

**Issue by khoin343** [Naveen3Singh/Simple-SMS-Messenger#123]
### Checklist

- [X] I can reproduce the bug with the latest version given [here](https://github.com/FossifyOrg/Messages/releases/latest).
- [X] I made sure that there are **no existing issues** - [open](https://github.com/FossifyOrg/Messages/issues) or [closed](https://github.com/FossifyOrg/Messages/issues?q=is%3Aissue+is%3Aclosed) - to which I could contribute my information.
- [X] I made sure that there are **no existing discussions** - [open](https://github.com/FossifyOrg/Messages/discussions) or [closed](https://github.com/FossifyOrg/Messages/discussions?discussions_q=is%3Aclosed) - to which I could contribute my information.
- [X] I have read the FAQs inside the app (Menu -> About -> FAQs) and my problem isn't listed.
- [X] **I have taken the time to fill in all the required details. I understand that the bug report will be dismissed otherwise.**
- [X] This issue contains only one bug.
- [X] I have read and understood the [contribution guidelines](https://github.com/FossifyOrg/Messages/blob/master/CONTRIBUTING.md).

### Affected app version

1.0.1

### Affected Android/Custom ROM version

Android 14/One UI 6

### Affected device model

Samsung Galaxy A14 5G

### How did you install the app?

F-Droid / IzzyOnDroid

### Steps to reproduce the bug

1. Receive an MMS message
2. Wait for the blank notification
3. Scroll around the thread until your message says "Sending..." and updates

### Expected behavior

Message smoothly appears as soon as it is received.

### Actual behavior

Notification appears but message does not in thread. It will only appear after a bit of scrolling.

**Comment by Aga-C** [comment]
While receiving an MMS message, are you already in that conversation, or are you opening it after receiving the notification?

**Comment by khoin343** [comment]
I am already in that conversation.

**Comment by Aga-C** [comment]
One more question - you wrote about scrolling around the thread. Isn't it enough to just scroll down? Maybe the problem isn't that messages take a long time to appear, but it's just that it doesn't scroll.

I reproduced it on my phone: got a notification and the thread was still showing the old message, but it was enough to just scroll down, and the message is there instantly.

**Comment by khoin343** [comment]
Funnily enough it fixed itself when I reinstalled it, but now it's incredibly laggy to type. That might be an issue for another time...

**Comment by khoin343** [comment]
Anddd now it's back. Great.

**Comment by khoin343** [comment]
Now when I try to scroll around after the blank notification it does nothing. I just have to wait a long time before it appears.

**Comment by Cwpute** [comment]
I notice that too, but for SMS *as well*.
I'm currently scrolling the very conversation i get a message in, receive a new message (that i get a notification for, but that's another subject), but said new message that could easily be displayed in the notification doesn't load for a couple of seconds - usually up to ~3 seconds, but it can sometimes take up to ~8 seconds!

There is no issue with my satellite coverage, my sim card hasn't been acting up (it's actually almost brand new), and my phone isn't old (Nokia TA-1157).

**Comment by Ten0** [comment]
I have a similar issue but AIUI when I click the notification it displays in the thread, but then if I go back to the menu and enter back the conversation the message does not appear anymore, and if I wait some time (at least more than 5 seconds but idk where precisely the limit is), then when I open the conversation, the message finally appears again.

App version 1.2.2

**Comment by naveensingh** [comment]
Do you guys have a lot of messages in those affected conversations?

**Comment by Ten0** [comment]
No.
This is the entire (scam) conversation I've had with the number here:

<img alt="ignoreImageMinify" src="https://github.com/user-attachments/assets/afbfff13-e6e4-4feb-9ab2-3c72dbbcd69b" width=184 />

For the last message, the first time I opened it it displayed properly, then I started reporting it so I opened the "report conversation" I have with 33700, french malicious SMS report service, then opened back that conversation and the message was gone, tried exiting and reopening and it still wouldn't appear, and then I reopened it maybe a minute later and the message was back.

**Comment by Ten0** [comment]
> Do you guys have a lot of messages in those affected conversations?

I consider this a different issue but it also seems that if I try to load old messages in conversations it's very slow as well (independently of whether the conversation itself has lots of messages).

I find it surprising that when inspecting the database structure I do not find messages to be indexed (by thread id, or sender phone number) and I wonder if that could be linked to that second issue.

<img alt="ignoreImageMinify" src="https://github.com/user-attachments/assets/58bd4913-1dd9-4468-a268-5724e739f650" width=600 />

It looks like that wouldn't help this to be fast:
https://github.com/FossifyOrg/Messages/blob/cfc16dd886847cc4dc186558c8a4f6cb442eee08/app/src/main/kotlin/org/fossify/messages/activities/ThreadActivity.kt#L754-L755

At first glance it looks like this would deserve an index on thread_id, date.

That would enable this query to run without having to go through all messages ever:
https://github.com/FossifyOrg/Messages/blob/cfc16dd886847cc4dc186558c8a4f6cb442eee08/app/src/main/kotlin/org/fossify/messages/extensions/Context.kt#L132-L135

**Comment by naveensingh** [comment]
> I consider this a different issue but it also seems that if I try to load old messages in conversations it's very slow as well (independently of whether the conversation itself has lots of messages).

Yeah, this issue doesn't seem to be performance-related. I have found that loading is quick enough on most devices, but indexing will definitely improve performance. There are many quick fixes in this codebase.

> https://github.com/FossifyOrg/Messages/blob/cfc16dd886847cc4dc186558c8a4f6cb442eee08/app/src/main/kotlin/org/fossify/messages/extensions/Context.kt#L132-L135

That extension function queries the system messages db, not the in-app db.

**Comment by Just-Regis** [comment]
Same issue with postponed messages.
Close and reopen app doesn't work.
You have to go to the conversation, scroll down to the postponed message sent for refreshing the main screen.
